lib/spandx/js/parsers/yarn.rb

```diff
@@ -0,0 +1,243 @@
+
+              break if next_token.value != indent
+
+              self.token = tokens.next
+            elsif prop_token.type == TOKEN_TYPES[:indent]
+              break if prop_token.value != indent
+
+              self.token = tokens.next
+            elsif prop_token.type == TOKEN_TYPES[:eof]
+              break
+            elsif prop_token.type == TOKEN_TYPES[:string]
+              key = prop_token.value
+
+              raise StandartError, 'Expected a key' if key.nil?
+
+              keys = [key]
+              self.token = tokens.next
+
+              #  support multiple keys
+              while token.type == TOKEN_TYPES[:comma]
+                self.token = tokens.next
+
+                raise StandardError, 'Expected string' if token.type != TOKEN_TYPES[:string]
+
+                keys << token.value
+                self.token = tokens.next
+              end
+
+              was_colon = token.type == TOKEN_TYPES[:colon]
+
+              self.token = tokens.next if was_colon
+
+              if valid_prop_value_token?(token)
+                keys.each do |k|
+                  obj[k] = token.value
+                end
+
+                self.token = tokens.next
+              elsif was_colon
+                # parse object
+                val = parse_tokenized_items(indent + 1)
+
+                keys.each do |k|
+                  obj[k] = val
+                end
+
+                break if indent.positive? && token.type != TOKEN_TYPES[:indent]
+
+              else
+                raise StandardError, 'Invalid value type'
+              end
+            elsif prop_token.type == TOKEN_TYPES[:comment]
+              self.token = tokens.next
+              next
+            else
+              raise StandardError, "Unkown token #{token}"
+            end
+          end
+
+          obj
+        end
+
+        def count_special_characters(input)
+          i = 0
+          i += 1 until [':', "\n", "\r", ',', ' '].include?(input[i]) || input[i].nil?
+
+          i
+        end
+
+        def count_indent_size(input)
+          indent_size = 1
+          indent_size += 1 while input[indent_size] == ' '
+
+          raise StandardError, 'Invalid number of spaces' if indent_size.odd?
+
+          indent_size
+        end
+
+        def extract_value(input)
+          between_quotes = /\"(.*?)\"/
+          val = input.match(between_quotes).to_s
+
+          [val.size, val]
+        end
+
+        def compatible?(yarn_file)
+          version_regex = /yarn lockfile v(#{LOCKFILE_VERSION})$/.freeze
+
+          lines = yarn_file.split("\n", 2) # get first two lines
+
+          lines.each do |line|
+            return true if line.match?(version_regex)
+          end
+
+          false
+        end
+
+        def convert_objects(dependencies)
+          deps = Set.new
+          dependencies.each do |dependency, details|
+            name = dependency.match(/(^.*)(@)(.*)/i).captures.first
```

Comment:
An alternative is to use `scan`. Also, named capture groups might make this a little more readable. E.g. (?name)